cclient/api/extended/prepare_transfers.c

```diff
@@ -5,27 +5,219 @@
-  retcode_t ret_code = RC_ERROR;
-  iota_transaction_t TX = {};
-  iota_transaction_t* tx = NULL;
-  transfer_iterator_t* transfer_iterator = NULL;
-  Kerl kerl = {};
+static void add_transaction_entry(bundle_transactions_t* bundle, int num_txs, flex_trit_t const* const address,
+                                  int64_t value, flex_trit_t* tag, uint64_t timestamp) {
+  iota_transaction_t tx = {};
+  for (int i = 0; i < num_txs; i++) {
+    transaction_reset(&tx);
+    transaction_set_address(&tx, address);
+    transaction_set_tag(&tx, tag);
+    transaction_set_timestamp(&tx, timestamp);
+    if (i == 0) {  // put value in first transaction only.
+      transaction_set_value(&tx, value);
+    } else {
+      transaction_set_value(&tx, 0);
+    }
+    transaction_set_obsolete_tag(&tx, tag);
+    bundle_transactions_add(bundle, &tx);
+  }
+}
 
-  transfer_iterator = transfer_iterator_new(transfers, num_transfer, &kerl, &TX);
+// add remainder and sign inputs
+static retcode_t add_remainder(iota_client_service_t const* const serv, flex_trit_t const* const seed, uint8_t security,
+                               inputs_t const* const inputs, bundle_transactions_t* const bundle, flex_trit_t* tag,
+                               int64_t total, flex_trit_t const* const remainder_address, Kerl* const kerl) {
+  retcode_t ret = RC_ERROR;
+  int64_t total_transfer_value = total;
+  input_t* input = NULL;
 
-  if (transfer_iterator) {
-    for (tx = transfer_iterator_next(transfer_iterator); tx; tx = transfer_iterator_next(transfer_iterator)) {
-      bundle_transactions_add(out_bundle, tx);
+  INPUTS_FOREACH(inputs->input_array, input) {
+    int64_t input_value = input->balance;
+    int64_t substract = 0 - input_value;
+    uint64_t timestamp = current_timestamp_ms();
+    if (input_value == 0) {
+      continue;
     }
-  } else {
-    ret_code = RC_CCLIENT_INVALID_TRANSFER;
-    log_error(client_extended_logger_id, "preparing transfer failed: %s\n", __func__, error_2_string(ret_code));
+
+    // add input to bundle
+    add_transaction_entry(bundle, security, input->address, substract, tag, timestamp);
+
+    if (input_value >= total_transfer_value) {
+      int64_t remainder_value = input_value - total_transfer_value;
+      if (remainder_value > 0 && remainder_address != NULL) {
+        // put remainder value to remainder address
+        add_transaction_entry(bundle, 1, remainder_address, remainder_value, tag, timestamp);
+        // sign and return
+        return bundle_sign(bundle, seed, inputs, kerl);
+      } else if (remainder_value > 0) {
+        // remainder address is needed
+        address_opt_t addr_info = {0, 0, security};
+        hash243_queue_t new_addresses = NULL;
+        if ((ret = iota_client_get_new_address(serv, seed, addr_info, &new_addresses)) == RC_OK) {
+          // unused address
+          add_transaction_entry(bundle, 1, new_addresses->prev->hash, remainder_value, tag, timestamp);
```

Comment:
Just confirming it's done on purpose: why `->prev->` ?

Reply:
it's the location of the unused address.

Reply:
Thanks